### PR TITLE
[shardformer] hybridparallelplugin support gradients accumulation.

### PR DESCRIPTION
**Pull request by flybird11111**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#5205 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

hybridparallelplugin support gradients accumulation.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: tests/test_booster/test_plugin/test_3d_grad_acc.py
Comment (ver217) [inline tests/test_booster/test_plugin/test_3d_grad_acc.py:1]: Move this to test_3d_plugin

Comment (flybird11111) [reply tests/test_booster/test_plugin/test_3d_grad_acc.py:1]: updated.

File: tests/test_booster/test_plugin/test_3d_plugin.py [hunk 1]
```diff
@@ -1,19 +1,46 @@
...
 from colossalai.testing import clear_cache_before_run, parameterize, rerun_if_address_is_in_use, spawn
+from colossalai.utils import get_current_device
 from tests.kit.model_zoo import model_zoo
 
 
+class RandomDataset(Dataset):
+    def __init__(self, num_samples: int = 100, max_length: int = 512, vocab_size: int = 32000):
+        self.num_samples = num_samples
+        self.max_length = max_length
+        self.input_ids = torch.randint(0, vocab_size // 2, (num_samples, max_length), device=get_current_device())
```
Comment (ver217) [inline tests/test_booster/test_plugin/test_3d_plugin.py:26]: Why `vocab_size // 2`

File: tests/test_booster/test_plugin/test_3d_plugin.py [hunk 2]
```diff
@@ -85,10 +112,149 @@ def check_3d_plugin(init_method: str = "none", early_stop: bool = True):
...
+    "test_args",
+    [
+        {
+            "batch_size": 2,
+            "num_steps": 8,
+            "tp": 2,
+            "pp": 2,
+            "pp_style": "1f1b",
+            "num_model_chunks": 1,
+            "mbs": 2,
```
Comment (ver217) [inline tests/test_booster/test_plugin/test_3d_plugin.py:125]: only 1 micro batch is not suitable

File: tests/test_booster/test_plugin/test_3d_plugin.py [hunk 3]
```diff
@@ -85,10 +112,149 @@ def check_3d_plugin(init_method: str = "none", early_stop: bool = True):
...
+            "zero": 0,
+            "precision": "fp16",
+            "initial_scale": 1,
+            "max_length": 512,
+            "gradient_accumulation_step": 4,
```
Comment (ver217) [inline tests/test_booster/test_plugin/test_3d_plugin.py:130]: Keep this small. 8 steps take too long

File: tests/test_booster/test_plugin/test_3d_plugin.py [hunk 4]
```diff
@@ -85,10 +112,149 @@ def check_3d_plugin(init_method: str = "none", early_stop: bool = True):
...
+        },
+    ],
+)
+def run_grad_acc_test(test_args):
+    _, (model_fn, _, _, _, _) = list(model_zoo.get_sub_registry("transformers_gpt_lm").items())[0]
```
Comment (ver217) [inline tests/test_booster/test_plugin/test_3d_plugin.py:177]: If you only want to test one model, you should use a more concrete name (not just "transformers_gpt_lm")

File: tests/test_booster/test_plugin/test_3d_plugin.py [hunk 5]
```diff
@@ -85,10 +112,149 @@ def check_3d_plugin(init_method: str = "none", early_stop: bool = True):
...
+    grad_accu_step = test_args["gradient_accumulation_step"]
+    for step, batch in enumerate(dataloader):
+        batch = move_to_cuda(batch)
+        # train origin model
+        origin_output = origin_model(**batch)
+        origin_loss = origin_output[0] / grad_accu_step
+        origin_loss.backward()
+
+        if (step + 1) % grad_accu_step != 0 and test_args["zero"] != 2:
+            sync_context = booster.no_sync(model, optimizer)
```
Comment (ver217) [inline tests/test_booster/test_plugin/test_3d_plugin.py:213]: the context should not be created in for-loop

File: tests/test_booster/test_plugin/test_3d_plugin.py [hunk 6]
```diff
@@ -85,10 +112,149 @@ def check_3d_plugin(init_method: str = "none", early_stop: bool = True):
...
+            optimizer.step()
+            optimizer.zero_grad()
+
+    if plugin.stage_manager is None:
+        for p1, p2 in zip(model.unwrap().parameters(), origin_model.parameters()):
+            if test_args["tp"] > 1:
+                p1 = gather_distributed_param(p1, keep_vars=False)
+            assert_close(p1.to(p2.dtype), p2, atol=1e-2, rtol=1e-2)
+    elif plugin.stage_manager.is_first_stage(ignore_chunk=True):
+        for p1, p2 in zip(model.unwrap().transformer.h[0].parameters(), origin_model.transformer.h[0].parameters()):
```
Comment (ver217) [inline tests/test_booster/test_plugin/test_3d_plugin.py:247]: This two conditions check different parameters, why?

File: tests/test_booster/test_plugin/test_3d_plugin.py [hunk 7]
```diff
@@ -85,10 +113,152 @@ def check_3d_plugin(init_method: str = "none", early_stop: bool = True):
...
+            origin_model, origin_optimizer, dataloader=dataloader
+        )
+        for p1, p2 in zip(model.unwrap().parameters(), origin_model.unwrap().parameters()):
+            assert_close(p1.to(p2.dtype), p2, atol=1e-2, rtol=1e-2)
+        return
+
+    for p1, p2 in zip(model.unwrap().parameters(), origin_model.parameters()):
+        if test_args["tp"] > 1:
+            p1 = gather_distributed_param(p1, keep_vars=False)
+        assert_close(p1.to(p2.dtype), p2, atol=1e-2, rtol=1e-2)
```
Comment (ver217) [inline tests/test_booster/test_plugin/test_3d_plugin.py:254]: Redundant?

**Review by ver217 — APPROVED** [review]
